### PR TITLE
Support reactor context span propagation

### PR DESCRIPTION
**Pull request by amarziali**
# What Does This Do

This PR add the possibility to manually add the current span in the reactor subscriber context using the key `dd.span` and allows addressing top-down context propagation.

Then, when the publisher has a span in its context will activate this span when it will signal downstream (`onNext`, `onError`, `onComplete`). That will make downstream operators using this span as active as well.

Here above few examples (see also unit tests on this PR for more):

1. No downstream context propagation

This is the basic example where context propagation happens only bottom up

```java
Span parent = buildSpan("parent").start();
try (Scope parentScope = activateSpan(parent)) {
Mono.just("Hello")
.flatMap(s ->  {
   final Span child = buildSpan("child").start();
   return Mono.just(s + " World).doFinally(ignored -> child.finish());
 }
. map(s -> {
   final Span last = buildSpan("last").start();
  try (Scope ignored = activateSpan(last)) {
     // do something interesting
    return s;
  } finally {
    last.finish();
 }
}).block()
parent.finish();
}
```

The scope of `parent` is propagated upstream when `block()` is called. The scope for `child` is not propagated upstream .

The produced trace looks like `[parent[child][last]`

2. Set active span on the chain

We now change slightly the previous example to let `child` propagating downstream when the flatmap emits.

```java
Span parent = buildSpan("parent").start();
try (Scope parentScope = activateSpan(parent)) {
Mono.just("Hello")
.flatMap(s ->  {
   final Span child = buildSpan("child").start();
   return Mono.just(s + " World).contextWrite("dd.span", child).doFinally(ignored -> child.finish());
 }
. map(s -> {
   final Span last = buildSpan("last").start();
  try (Scope ignored = activateSpan(last)) {
     // do something interesting
    return s;
  } finally {
    last.finish();
 }
}).block()
parent.finish();
}
```

When we indicate that we want to emit `child` as active span, the map operator will see now that span and not the `parent` span captured when `subscribe` was called

Hence the trace structure is now `[parent[child[last]]]`

3. Restore a span in the chain

Previously, forcing an operator to emit a different active span caused all the downstream chain to see that span as active.
We might want to change the active span on the way down. The example will be now:

```java
Span parent = buildSpan("parent").start();
try (Scope parentScope = activateSpan(parent)) {
Mono.just("Hello")
.flatMap(s ->  {
   final Span child = buildSpan("child").start();
   return Mono.just(s + " World)
   // do something more with child as active span in the flatmap ...
   // ....
   .contextWrite("dd.span", child).doFinally(ignored -> child.finish());
 }
// now change back to parent
.contextWrite("dd.span", parent)
. map(s -> {
   final Span last = buildSpan("last").start();
  try (Scope ignored = activateSpan(last)) {
     // do something interesting
    return s;
  } finally {
    last.finish();
 }
}).block()
parent.finish();
}
```

Each time we use `contextWrite` setting `dd.span` it will propagate upstream that information so we should carefully set and reset that information in the right places

The trace structure should look like `[parent[child][last]]`
# Motivation

# Additional Notes

# Contributor Checklist

- [ ] Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- [ ] Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- [ ] Squash your commits prior merging or merge using GitHub's [Squash and merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits)
- [ ] Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- [ ] Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [PROJ-IDENT]

### CODE REVIEW

File: dd-java-agent/instrumentation/reactor-core-3.1/src/main/java/datadog/trace/instrumentation/reactor/core/SpanExtractorHelper.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    }
+  }
+
+  private static final ClassValue<Function<?, AgentSpan>> DELEGATE_CLASS_VALUE =
+      GenericClassValue.of(
+          aClass -> {
+            if (AgentSpan.class.isAssignableFrom(aClass)) {
+              return Function.identity();
+            } else if (MutableSpan.class.isAssignableFrom(aClass)
+                || "datadog.opentelemetry.shim.trace.OtelSpan".equals(aClass.getName())) {
```
[the remark this answers is not in the document]
Comment (amarziali) [reply dd-java-agent/instrumentation/reactor-core-3.1/src/main/java/datadog/trace/instrumentation/reactor/core/SpanExtractorHelper.java]: I introduced `WithAgentSpan` to have the original span when wrapped in other structures (I did a little bit more refactoring hence)

**Review by mcculls — APPROVED** [review]
Nice refactoring, just a minor comment about re-using the value of `self.currentContext()`

File: internal-api/src/main/java/datadog/trace/bootstrap/instrumentation/api/WithAgentSpan.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package datadog.trace.bootstrap.instrumentation.api;
+
+public interface WithAgentSpan {
```
Comment (amarziali) [inline internal-api/src/main/java/datadog/trace/bootstrap/instrumentation/api/WithAgentSpan.java:3]: Note for future improvements: we might expose the behaviour (i.e. `activate()`) instead the internal AgentSpan object

**Review by PerfectSlayer — COMMENTED** [review]
As future design idea, I would like to have the OT/OTel span capabilities exposed rather than their internals (delegate span).

For example, having an `ActivableSpan` interface that exposes the `activate()` method rather than `WithAgentSpan` that exposes the `delegate` internal.

```java
      if (context.hasKey("dd.span")) {
        Object maybeSpan = context.get("dd.span");
        if (maybeSpan instanceof WithAgentSpan) {
          AgentSpan span = ((WithAgentSpan) maybeSpan).asAgentSpan();
          if (span != null) {
            return activateSpan(span);
          }
        }
      }
```
will become:
```java
      if (context.hasKey("dd.span")) {
        Object maybeSpan = context.get("dd.span");
        if (maybeSpan instanceof ActivableSpan) {
          return ((ActivableSpan) maybeSpan).activate();
        }
      }
```

Even though OT/OTel spans should rely on `AgentSpan` for a long time, it feels safer to expose capability rather than internal. That's not blocking though :)

**Pull request merged**